src/utils/return_periods.py

```diff
@@ -1,8 +1,109 @@
 import numpy as np
 import pandas as pd
 from lmoments3 import distr, lmom_ratios
+from scipy.interpolate import interp1d
 from scipy.stats import norm, pearson3, skew
 
+# Empirical RP Functions
+
+
+def fs_add_rp(df, df_maxima, by):
+    """
+    Adds return periods (RP) to the given DataFrame based on maxima values.
+    from entire historical record
+
+    Parameters:
+    df (pandas.DataFrame): The DataFrame to which return periods will be added.
+    df_maxima (pandas.DataFrame): The DataFrame containing maxima values
+    from entire historical record - used to calculate return periods.
+    by (str or list of str): Column(s) to group by when calculating
+    return periods.
+
+    Returns:
+    pandas.DataFrame: The original DataFrame with an additional column 'RP'
+    containing the return periods.
+    """
+    df_maxima["RP"] = df_maxima.groupby(by)["value"].transform(empirical_rp)
+
+    interp_funcs = interpolation_functions_by(
+        df=df_maxima, rp="RP", value="value", by=by
+    )
+
+    df["RP"] = df.apply(
+        lambda row: apply_interp(row, interp_funcs, by=by), axis=1
+    ).astype(float)
+
+    df["RP"] = clean_rps(df["RP"], decimals=3, upper=10)
+
+    return df
+
+
+def interpolation_functions_by(df, rp, value, by=["iso3", "pcode"]):
+    """
+    Generate interpolation functions for each group in the DataFrame.
+
+    Parameters:
+    df (pandas.DataFrame): The input DataFrame containing the data.
+    rp (str): The column name in the DataFrame representing the
+    return period values.
+    value (str): The column name in the DataFrame representing the values
+    to interpolate.
+    by (list of str, optional): The list of column names to group by.
+    Default is ["iso3", "pcode"].
+
+    Returns:
+    dict: A dictionary where keys are tuples of group values and values are
+    interpolation functions.
+    """
+    interp_funcs = (
+        df.groupby(by)
+        .apply(
+            lambda group: interp1d(
```

Comment:
Hmm I was just thinking whether linear interpolation makes the most sense here, or whether we should take the log and interpolate against this (i.e., "geometric" interpolation instead of arithmetic). I think in the relatively small range  of RPs here, linear is fine (and this is what I've always used), but since we don't expect the RP-RV curve to be linear, we might not want to use a linear interpolation.

All this being said- in most cases, in practice it won't make much of a difference.

Reply:
what function would you think? I'm having a hard time visualizing a non-linear interpolation line that wouldn't introduce some weird artefacts/effects ?